XCM overview (#157)
moondance-labs/tanssi-docs: learn/framework/xcm.md

## Common Use Cases {: #common-use-cases }

Many use cases can be addressed by benefiting from the common ground and versatility XCM provides. Two of the most recurrent ones are asset transfers and remote execution.

### Asset Transfers {: #asset-transfer }

Moving digital assets from one appchain to another is essential for creating a more dynamic, efficient, and interconnected blockchain ecosystem. The native cross-chain capability allows two main strategies to transfer assets from one chain to another:

- **Teleport** - teleporting an asset is a simple and efficient mechanism, but it has a major caveat: it requires trust between the parties. In essence, when appchain A wants to send X amount of assets to appchain B, it burns X amount of assets and sends a message to appchain B instructing them to mint exactly X amount of assets, preserving the overall asset balance and concluding the teleport action. In this process, appchain A trusts appchain B not to mint more tokens than what was transferred, and appchain B trusts appchain A to burn the tokens that were transferred

> - **Reserve transfer** - A reserve transfer involves the **reserve chain** of an asset, which is the chain where the asset is native (e.g., [Moonbeam](https://moonbeam.network/){target=\_blank} is the reserve chain for the GLMR token). Also, non-reserve appchains hold a **sovereign account** on the reserve chain, a keyless account managed by the respective appchain governor. Thus, when reserve appchain A wants to send X amount of an asset to non-reserve appchain B, it locally transfers the assets to appchain's B sovereign account and, in the same atomic action, it sends an XCM message to appchain B with instructions to mint X amount of a derivative form of the transferred asset. On the other way around, if non-reserve appchain B wants to send X amount of an asset to reserve appchain A, then the steps are: appchain B burns the derived form of the asset locally and sends an XCM message to appchain A, with instructions to transfer the assets from appchain B's sovereign account to appchain's A destination account. Even if the non-reserve appchain mints derived tokens in excess (or doesn't burn tokens when transferring), these tokens will have no real value because they are not backed one-to-one in the reserve chain

Reviewer: ```suggestion
- **Reserve transfer** - A reserve transfer involves the **reserve chain** of an asset, which is the chain where the asset is native (e.g., [Moonbeam](https://moonbeam.network/){target=\_blank} is the reserve chain for the GLMR token). Also, non-reserve appchains hold a *sovereign account* on the reserve chain, a keyless account managed by the respective appchain governor. Thus, when reserve appchain A wants to send X amount of an asset to non-reserve appchain B, it locally transfers the assets to appchain's B sovereign account and, in the same atomic action, it sends an XCM message to appchain B with instructions to mint X amount of a derivative form of the transferred asset. On the other way around, if non-reserve appchain B wants to send X amount of an asset to reserve appchain A, then the steps are: appchain B burns the derived form of the asset locally and sends an XCM message to appchain A, with instructions to transfer the assets from appchain B's sovereign account to appchain's A destination account. Even if the non-reserve appchain mints derived tokens in excess (or doesn't burn tokens when transferring), these tokens will have no real value because they are not backed one-to-one in the reserve chain
```

Let's go italics with this one. https://developers.google.com/style/key-terms

The associated fees for executing transfers are typically deducted from the transferred amount, so the recipient receives the intended amount minus the fees.

### Remote Execution {: #remote-execution }

The native interoperability XCM provides allows an appchain to send a message to another triggering some action. For example, If the destination chain is EVM-compatible, appchain A can call a smart contract deployed on appchain B.

As mentioned in the [fees section](#fees), to get any on-chain request executed it is necessary to cover its associated fees. On XCM, remote execution can be bought with two steps:

> 1. Reserve some assets using the *WithdrawAsset* XCM instruction, which takes funds from the call origin and puts them in a holding register
> 2. Pay for the on-chain execution, using the *BuyExecution* XCM instruction, which uses the previously withdrawn assets

Reviewer: ```suggestion
1. Reserve some assets using the `WithdrawAsset` XCM instruction, which takes funds from the call origin and puts them in a holding register
2. Pay for the on-chain execution, using the `BuyExecution` XCM instruction, which uses the previously withdrawn assets
```

!!! note

>     When an appchain sends an XCM message, its default source on the receiving end is the origin appchain's Sovereign account. The sender appchain can add an XCM instruction called *DescendOrigin* to the message, changing the origin account to match the signing user's account, ensuring execution occurs on behalf of the same entity initiating the XCM message on the source chain, and avoiding a potentially unsafe scenario.

Reviewer: ```suggestion
    When an appchain sends an XCM message, its default source on the receiving end is the origin appchain's Sovereign account. The sender appchain can add an XCM instruction called `DescendOrigin` to the message, changing the origin account to match the signing user's account, ensuring execution occurs on behalf of the same entity initiating the XCM message on the source chain, and avoiding a potentially unsafe scenario.
```

> Finally, the execution takes place on the destination chain, calling a smart contract or any other transaction using the XCM instruction called *Transact*.

Reviewer: ```suggestion
Finally, the execution takes place on the destination chain, calling a smart contract or any other transaction using the XCM instruction called `Transact`.
```

## Establishing Cross-Chain Communication {: #channel-registration }

Before two chains can communicate, a messaging channel must be established. Channels are unidirectional, which means that separate channels are needed to send messages from chain A to chain B and B to A.

For chain A to communicate with chain B, chain A must send an open channel transaction to the relay chain requesting a channel be opened with chain B. Chain B must then accept the request by sending a corresponding XCM message to the relay chain. Only when both chains agree is the channel opened in the next epoch. The same process is required to establish a channel from chain B to chain A.

It is important to note that a channel between an appchain and the relay chain is automatically opened upon appchain registration and onboarding.

![XCM Channel Registration Overview](/images/learn/framework/xcm/dark-xcm-2.webp#only-dark)
![XCM Channel Registration Overview](/images/learn/framework/xcm/light-xcm-2.webp#only-dark#only-light)

Once the channel is established, cross-chain messages can be sent between appchains. For asset transfers, assets will also need to be registered before being transferred.

!!! note

>     XCM is a versioned ever-evolving language. When two communicating appchains are using different XCM versions, they must use the latest version supported by the less upgraded side. To find out the latest XCM version an appchain can work with, other appchains can query it and subscribe for updates whenever this changes.

Reviewer: ```suggestion
    XCM is a versioned, ever-evolving language. When two communicating appchains use different XCM versions, they must use the latest version supported by the less upgraded side. To find out the latest XCM version an appchain can work with, other appchains can query it and subscribe for updates whenever this changes.
```

To compose meaningful messages in a multichain environment it is necessary to have a precise yet abstract way of referencing resources located in different consensus systems. A concept called *multilocation* is used to serve this purpose and target a specific chain or any of its inner elements, such as an account, an asset, or a smart contract.

XCM's destination elements are organized in a hierarchical architecture, where elements are contained within other components. For example, a smart contract is an element contained within an appchain, and the same can be said for an account or an ERC20 asset. Appchains are contained by the relay chain, which plays a crucial role in the cross-chain messaging process, relaying messages from one appchain to another.

Multilocations are not a universal resource locator. They refer to elements from the sender's perspective and are composed of two components: `parents` and `interior`. Parents is a property that indicates if the route must "move up" in the hierarchy, i.e., from an appchain to the relay chain. Interior is a list of junctions that define how to locate the destination. Here are some examples of multilocations:

> - **Appchain A references a smart contract in appchain B** - from the point of view of appchain A, to reach a smart contract in appchain B it is necessary to move up in the hierarchy (to the relay chain) and then descend to appchain B to, once there, reference the smart contract's address. The *multilocation* is therefore defined with a `parents` value set to `1`, which moves up, and two junctions, one defining which appchain should receive the message, and the other defining the H160 address of the smart contract that will be called

Reviewer: ```suggestion
- **Appchain A references a smart contract in appchain B** - from the point of view of appchain A, to reach a smart contract in appchain B it is necessary to move up in the hierarchy (to the relay chain) and then descend to appchain B to, once there, reference the smart contract's address. The multilocation is therefore defined with a `parents` value set to `1`, which moves up, and two junctions, one defining which appchain should receive the message, and the other defining the H160 address of the smart contract that will be called
```

No need to continue to put in italics after it has been introduced

> ![Smart Contract Multilocation Example](/images/learn/framework/xcm/dark-xcm-3.webp#only-dark)

Reviewer: Are these using papermoon colors?!

![Smart Contract Multilocation Example](/images/learn/framework/xcm/light-xcm-3.webp#only-light)

> - **Appchain A references an account in the relay chain** - from the point of view of appchain A, to reference an account in the relay chain, it is necessary to move up and then reference the account. The *multilocation* is defined with a `parents` value set to `1`, which moves up to the relay chain, and one junction that references the substrate type destination address 

Reviewer: ```suggestion
- **Appchain A references an account in the relay chain** - from the point of view of appchain A, to reference an account in the relay chain, it is necessary to move up and then reference the account. The multilocation is defined with a `parents` value set to `1`, which moves up to the relay chain, and one junction that references the substrate type destination address 
```

---
title: Native Cross-Chain Communication
description: Tanssi appchains benefit from XCM, a native cross-chain communication language, which allows fast and secure bridging guaranteed by Polkadot's relay chain.
---

# Native Cross-Chain Communication

## Introduction {: #introduction }

> All the Tanssi appchains have an inherent capability to communicate and interoperate with any other appchain from the Polkadot ecosystem (be it Tanssi-powered or not). This native cross-chain communication feature is possible thanks to the Polkadot relay chain, which guarantees message delivery, and the XCM (Cross-Consensus Message) format, which provides a lingua franca.

Reviewer 2: Find a synonym for "lingua franca" feels weird haha.

Also - I would not specifically say "thanks to Polkadot relay chain" to move away from the Polkadot Centric messaging, but more something like "is possible thanks to the architecture Tanssi appchains are built on top off"

> All the Tanssi appchains have an inherent capability to communicate and interoperate with any other appchain from the Polkadot ecosystem (be it Tanssi-powered or not). This native cross-chain communication feature is possible thanks to the Polkadot relay chain, which guarantees message delivery, and the XCM (Cross-Consensus Message) format, which provides a lingua franca.

Author: hahah word of the day :)

> XCM is a messaging language designed to be generic. It doesn't make any assumptions about the destination chain and can communicate different intentions between sovereign consensus systems. An XCM message is, in fact, a program holding one or more instructions that will be relayed for execution to the destination chain.

Reviewer 2: ```suggestion
XCM is a messaging language designed to be generic. It doesn't make any assumptions about the destination chain and can communicate different intentions between sovereign consensus systems.

An XCM message is a program holding one or more instructions that will be relayed for execution to the destination chain. By itself, each XCM instruction is meaningless, but the combination of a specific set of instructions can result in a desired action when the XCM message is executed in the destination chain.
```

- **Absolute** -  messages are guaranteed to be delivered to the intended destination, in order and in a timely fashion
- **Asymmetric** -  messages sent have no response counterpart. Any return values, if required, must be sent back from the destination to the sender with another message
- **Agnostic** -  there are no assumptions whatsoever about the configuration or properties of the appchains that are communicating. Two communicating appchains might differ in every aspect, except the ability to understand XCM. E.g., one chain could be EVM-compatible and the other not, one chain could be a DeFi appchain and the other a gaming appchain, etc

## Common Use Cases {: #common-use-cases }

Benefiting from the common ground and versatility XCM provides, many use cases can be addressed. Two of the most recurrent ones are asset transfers and remote execution.

### Asset Transfers {: #asset-transfer }

Moving digital assets from one appchain to another is essential for creating a more dynamic, efficient, and interconnected blockchain ecosystem. The native cross-chain capability allows two main strategies to transfer assets from one chain to another:

> - **Teleport** - teleporting an asset is a simple and efficient mechanism, but it has a major caveat: it requires trust between the parties. In essence, when appchain A wants to send X amount of the assets to appchain B, it sends a message communicating appchain B that they must mint exactly X amount of the assets. After relaying the message, appchain A burns X amount of the assets, keeping the overall asset balance and concluding the teleport action

Reviewer 2: In teleport from Chain A to Chain B

Chain A burns the assets and in the same action sends the message to Chain B that should mint whatever Chain A burned, and then set apart some for fees (if that is the case)

> - **Reserve transfer** - a reserve transfer involves a trusted third party acting as a reserve of value. When appchain A wants to send X amount of an asset to appchain B, it sends the instruction to the reserve, who transfers the assets from appchain A to appchain B reserve accounts. The process is completed when appchain A burns X amount of the assets and appchain B mints them. Note that regardless of appchain A not burning the assets or appchain B minting tokens in excess, since the wrongful tokens are not backed one-to-one by the reserve, they have no real value

Reviewer 2: So in reserve transfer there is no trusted third party. Just that you don't need to trust the other chain involved in the process, because total value (or amount of tokens) are kept constant in the origin chain. 

What happens in that the destination chain owns an keyless account in the origin chain were funds are deposit to through a specific XCM instruction. That XCM instructions at the same time sends an XCM to that destination chain specific to minting that assets that are now held in their sovereign account in the origin chain

Reviewer 2: here is also important the concept of reserve. What is the reserve chain? The chain where the asset actually originates.

Let's discuss this in Tomorrow's learning session

### Remote Execution {: #remote-execution }

> Another common use is remote execution, which is sending a message to an appchain and triggering some action. If the destination chain is an EVM-compatible one, this allows appchain A to call a smart contract deployed on appchain B, for example.

Reviewer 2: I would expand this a bit. Expand on what is the general flow.

Some thing are similar between asset transfer and remote execution that you never touch upon. For example, the need of XCM delivery fees (which should be cover here) and also the need of buying execution in the destination chain

## Establishing Cross-Chain Communication {: #channel-registration }

Before two chains can start communicating, a messaging channel must be opened. Channels are unidirectional, meaning a channel from chain A to chain B will only pass messages from A to B. Therefore, two channels must be opened to send messages back and forth.

A channel between the appchain and the relay chain is automatically opened upon appchain registration and onboarding. However, when appchain A wants to open a communication channel with appchain B, appchain A must send an open channel transaction to its network. This transaction is an XCM message as well!

It's important to note that appchain A expressing its intentions of opening an XCM channel with appchain B is not sufficient. Appchain B must also accept the request by sending another XCM message. Only when both appchains have reached a mutual agreement, the channel is opened within the following epoch. The same process must be repeated to open a channel from appchain B to appchain A.

> ![XCM Channel Registration Overview](/images/learn/framework/xcm/dark-xcm-1.webp#only-dark)

Reviewer 2: Let's make images with a bit more Tanssi branding. IN the paragraph above you also never mention that the XCM message to open and accept channels are actually sent to the relay chain!

![XCM Channel Registration Overview](/images/learn/framework/xcm/light-xcm-1.webp#only-dark#only-light)

> Once the channel is established, cross-chain messages can be sent between appchains. For asset transfers, assets will also need to be registered before being transferred.

Reviewer 2: We never mention here also that the first XCM message sent between networks is to subscribe to the XCM version being used by the other chain. So that when the XCM version changes all the chains that there are channels opened should get notified of the version change

## Message Destinations {: #message-destinations }

> Any given appchain contains modules exposing behavior through transactions or smart contracts that can be called, accounts with balances, and assets (fungibles or non-fungibles) that can be created, transferred, etc. Also, all the appchains within the ecosystem are L1 blockchains connected to an L0 blockchain, the relay chain, which plays a crucial role in the cross-chain messaging process, with the ability to relay messages from one appchain to another.

Reviewer 2: Don't use `etc` it is better to use `among other` or similar things.

> To send and receive cross-chain messages that accurately reference the destination chain, its exposed behavior (such as smart contracts or transactions), and its available assets, a mechanism called *Multilocation* is used. Multilocations describe an element from the sender's perspective, so they are not intended to provide a universal description of the element.

Reviewer 2: I don't really like how you get to the point of stating Multilocaitons. I would first phrase a problem. How do you define something in an ecosystem that is natively cross-chain?

Multilocations are not mechanisms are a way to define anything in a Tanssi-powered Appchain. You can define even Ethereum or a smart contract on Ethereum!

> Two components define a location: `parents` and `interior`. Parents is a value that indicates if the route must move up to the relay chain. Interior is a list of junctions, that define how to locate the destination. Let's list some examples:

Reviewer 2: It is not really up to the relay chain; that is a way that I normally use because it is easier to understand. Unsure what is the best way here

> Two components define a location: `parents` and `interior`. Parents is a value that indicates if the route must move up to the relay chain. Interior is a list of junctions, that define how to locate the destination. Let's list some examples:

Reviewer: Might be worth noting here before diving into the examples that multilocations are hierarchical structures. So that is why an "upward" move is necessary to get from a parachain to the relay chain

I prefer not to use "let's"

Reviewer: Love this examples section though! 😍

> - **Appchain A references a smart contract in appchain B** - from the point of view of appchain A, to reach a smart contract in appchain B it is necessary to move up to the relay chain and then descend to appchain B and, once there, reference the smart contract's address. The *multilocation* is defined with a `parents` value set to 1, which moves up to the relay chain, and two junctions, one defining which appchain should receive the message, and the other defining the H160 address of the smart contract that will be called

Reviewer 2: Would love to see an image here to define this. I have ones related to multilocations

> - **Appchain A references a smart contract in appchain B** - from the point of view of appchain A, to reach a smart contract in appchain B it is necessary to move up to the relay chain and then descend to appchain B and, once there, reference the smart contract's address. The *multilocation* is defined with a `parents` value set to 1, which moves up to the relay chain, and two junctions, one defining which appchain should receive the message, and the other defining the H160 address of the smart contract that will be called

Reviewer: put the values in back ticks - so `1`

- **Appchain A references an account in the relay chain** - from the point of view of appchain A, to reference an account in the relay chain, it is necessary to move up and then reference the account. The *multilocation* is defined with a `parents` value set to 1, which moves up to the relay chain, and one junction that references the substrate type destination address 

## Fees {: #fees }

> A user executing a transaction on an appchain must pay the fees derived from computational effort associated with the task, and cross-chain execution is no exception to this rule. In cross-chain communication, a message requires execution on at least two different chains, and the user needs to pay for the fees associated with the computational effort made by every chain involved.

Reviewer 2: I think this should be way higher in the page

> For example, if a user in appchain A wants to call a smart contract in appchain B, the user must include in the XCM message the instructions to reserve a valid asset to cover the associated fees (being a valid asset, a token that is accepted by appchain B as payment for its services). With this valid asset provided, now the execution can be bought on the destination chain. 

Reviewer 2: Reserve a valid asset? I don't really like this phrasing. I think the right way would be to, user must include in the XCM message a way to provide a valid asset that can be used to pay for XCM execution in the destination chain, to cover the associated fees.

> For example, if a user on appchain A wants to call a smart contract on appchain B, the user must include instructions in the XCM message to reserve a valid asset to cover the associated fees. A valid asset is a token that appchain B accepts as payment for its services. Once this valid asset is provided, the execution can now be bought on the destination chain.

Reviewer: dupe

- **Absolute** -  messages are guaranteed to be delivered to the intended destination, in order and in a timely fashion
- **Asymmetric** -  messages sent have no response counterpart. Any return values, if required, must be sent back from the destination to the sender with another message
- **Agnostic** -  there are no assumptions whatsoever about the configuration or properties of two communicating appchains. Appchains might differ in every aspect, except the ability to understand XCM. E.g., one chain could be EVM-compatible and not the other, one chain could be a DeFi appchain and the other a gaming appchain, and so on

## Common Use Cases {: #common-use-cases }

Many use cases can be addressed by benefiting from the common ground and versatility XCM provides. Two of the most recurrent ones are asset transfers and remote execution.

### Asset Transfers {: #asset-transfer }

Moving digital assets from one appchain to another is essential for creating a more dynamic, efficient, and interconnected blockchain ecosystem. The native cross-chain capability allows two main strategies to transfer assets from one chain to another:

> - **Teleport** - teleporting an asset is a simple and efficient mechanism, but it has a major caveat: it requires trust between the parties. In essence, when appchain A wants to send X amount of assets to appchain B, it burns X amount of assets and sends a message to appchain B instructing them to mint exactly X amount of assets, preserving the overall asset balance and concluding the teleport action

Reviewer 2: Not really required but maybe we can explain why it requires "trust"

> - **Reserve transfer** - A reserve transfer is executed on the **reserve chain**, which is the chain where the asset being transferred is native (e.g., [Moonbeam](https://moonbeam.network/){target=\_blank} is the reserve chain for the GLMR token). Also, appchains hold a **sovereign account** on the reserve chain, a keyless account managed by the respective appchain governor. Thus, when appchain A wants to send X amount of an asset to appchain B, it sends the instructions to the reserve chain, which transfers the assets from appchain A to appchain B's sovereign accounts, and the instructions to appchain B to mint X amount of a derivative form of the transferred asset. Note that regardless of appchain B minting tokens in excess, since the wrongful tokens are not backed one-to-one by the reserve, they have no real value

Reviewer 2: It is not necessarily executed on the reserve chain, but it is generally between an asset reserve chain and a non-reserve chain. It can be executed on either

Reviewer 2: ```suggestion
- **Reserve transfer** - A reserve transfer involves the reserve chain of an asset, which is the chain where the asset being transferred is native (e.g., [Moonbeam](https://moonbeam.network/){target=\_blank} is the reserve chain for the GLMR token). Also, non-reserve appchains hold a **sovereign account** on the reserve chain, a keyless account managed by the respective appchain governor. Thus, when reserve appchain A wants to send X amount of an asset to non-reserve appchain B, it transfers the assets to appchain B's sovereign account in appchain A. In the same atomic action, it sends an XCM to appchain B with instructions to mint X amount of a derivative form of the transferred asset. Even if appchain B mints an excess of tokens, these tokens have no real value because they are not backed one-to-one in the reserve chain.
```

### Remote Execution {: #remote-execution }

The native interoperability XCM provides allows an appchain to send a message to another triggering some action. For example, If the destination chain is EVM-compatible, appchain A can call a smart contract deployed on appchain B.

> To get any on-chain request executed, it is necessary to cover its associated fees. On XCM, remote execution can be bought with two steps:

Reviewer 2: This is also true for token transfers, so maybe move this to the general chapter. XCM execution in  the destination chain also neds to buy execution

Author: Moved the fees section up, so it comes before remote transfers and execution.
(Also added a line about fees in the transfer section)
